src/FindIntrinsics.cpp

```diff
@@ -0,0 +1,714 @@
+            return IRMutator::visit(op);
+        }
+
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+
+        // Try widening both from the same type as the result, and from uint.
+        for (halide_type_code_t code : {op->type.code(), halide_type_uint}) {
+            Type narrow = op->type.narrow().with_code(code);
+            Expr narrow_a = lossless_cast(narrow, a);
+            Expr narrow_b = lossless_cast(narrow, b);
+
+            if (narrow_a.defined() && narrow_b.defined()) {
+                Expr negative_narrow_b = lossless_negate(narrow_b);
+                Expr result;
+                if (negative_narrow_b.defined()) {
+                    result = widening_add(narrow_a, negative_narrow_b);
+                } else {
+                    result = widening_sub(narrow_a, narrow_b);
+                }
+                if (result.type() != op->type) {
+                    result = Cast::make(op->type, result);
+                }
+                return mutate(result);
+            }
+        }
+
+        Expr negative_b = lossless_negate(b);
+        if (negative_b.defined()) {
+            return Add::make(a, negative_b);
+        }
+
+        if (a.same_as(op->a) && b.same_as(op->b)) {
+            return op;
+        } else {
+            return Sub::make(a, b);
+        }
+    }
+
+    Expr visit(const Mul *op) override {
+        if (!find_intrinsics_for_type(op->type)) {
+            return IRMutator::visit(op);
+        }
+
+        if (as_const_int(op->b) || as_const_uint(op->b)) {
+            // Distribute constants through add/sub. Do this before we muck everything up with widening
+            // intrinsics.
+            // TODO: Only do this for widening?
+            // TODO: Try to do this with IRMatcher::rewriter. The challenge is managing the narrowing/widening casts,
+            // and doing constant folding without the simplifier undoing the work.
+            if (const Add *add_a = op->a.as<Add>()) {
+                return mutate(Add::make(simplify(Mul::make(add_a->a, op->b)), simplify(Mul::make(add_a->b, op->b))));
+            } else if (const Sub *sub_a = op->a.as<Sub>()) {
+                return mutate(Sub::make(simplify(Mul::make(sub_a->a, op->b)), simplify(Mul::make(sub_a->b, op->b))));
+            }
+        }
+
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+
+        // Rewrite multiplies to shifts if possible.
+        if (op->type.is_int() || op->type.is_uint()) {
+            int pow2 = 0;
+            if (is_const_power_of_two_integer(a, &pow2)) {
+                return mutate(b << cast(UInt(b.type().bits()), pow2));
+            } else if (is_const_power_of_two_integer(b, &pow2)) {
+                return mutate(a << cast(UInt(a.type().bits()), pow2));
+            }
+        }
+
+        // We're applying this to float, which seems OK? float16 * float16 -> float32 is a widening multiply?
+        Expr narrow_a = strip_widening_cast(a);
```

Comment:
What's the difference between strip_widening_cast and lossless_narrow?

Reply:
strip_widening_cast removes all widening casts, ignoring sign and including more than narrowing (e.g. double narrowing).

Reply:
Oh, I guess it doesn't handle double narrowing. But it does ignore sign (the implementation is just calling lossless_cast but with some logic for signedness).